### ISSUE TITLE
'Fan only' missing from thermostatOperatingState

### ISSUE THREAD

**Issue by gaboferrari** [tonesto7/nest-manager#445]
## Verify the following before opening an issue:
- [x] That _OAuth_ is Enabled for the SmartApp under the IDE.
- [x] The SmartApp and Device are using the latest version.
- [x] You have reviewed the Readme for potential fixes.
- [x] That the Device Types have been _Published for You_.

## About Your Setup:
* How many Thermostats?: 1
* How many Protects?: 0
* Are you Using the Nest Presence Device?: No
* iOS or Android?: Android
* Mobile App Version: 2.3.1

## Expected Behavior
Change/Improvement
Device Handler for Nest Thermostat shall report "fan only" under thermostatOperatingState. The expected behavior is that the device handler knows and shares that state (fan only) to other smartapps. 

## Current Behavior

Current Device Handler for Nest Thermostat doesn't report "fan only" under thermostatOperatingState and shall report it so it can be used by smartapps.

## Steps to Reproduce (for bugs)
1.
2.
3.
4.

## Context
I am trying to use yvesracine's ScheduleTstatZones SmartApp to control my keen vents using Nest Thermostat and NST Manager. One of the option in yvesracine's app uses "fan only" state to open all Vents to 100%. Due to the missing state in the NST Manager/Nest Thermostat DH, I am not able to use that feature.

**Comment by imnotbob** [comment]
In smartthings, the fan is reported via http://docs.smartthings.com/en/latest/capabilities-reference.html#thermostat-fan-mode

Nest does comply with this reporting / setting.

In ST, thermostatmode does not report this as it is not part of the ST capabilities for thermostat mode:

http://docs.smartthings.com/en/latest/capabilities-reference.html#thermostat-mode

I'm sure yveses is aware of all this.

**Comment by yracine** [comment]
Hi, 

I asked Gabriel to submit this issue as the NST Manager is not following the ST standard capabilities which is causing some issues for all _ScheduleTstatZones_ users who are using this DTH and want their vents open when the thermostatOperatingState is 'fan only'.

Here are the standard events for the thermstatOperatingState attribute as indicated in  the SmartThings documentation:

http://docs.smartthings.com/en/latest/capabilities-reference.html#thermostat-operating-state

Also, there are other issues related to the way the NST manager is handling its setpoints.  Both cooling and heating setpoints should always be populated.  The NST manager should not obfuscate any setpoints regardless of the thermostat mode.

I understand that this issue a separate issue, but this is causing people to contact me for support when the problem is the DTH.

The setpoints handling is causing issues to other smartapps as well:

https://community.smartthings.com/t/beta-release-keenect-v2-7-and-keenzone-2-8-proportional-vent-control-with-system-wide-zone-balancing/59739/131

Regards.

**Comment by gaboferrari** [comment]
My request is not related to thermostatMode nor thermostatFanMode.

My request is related to thermostatOperatingState Capability. See below for the ST Documentation and other DH examples that contain this capability.

![image](https://cloud.githubusercontent.com/assets/18097540/25669336/34c58534-2ff7-11e7-9d67-5d12309068e9.png)

SmartThingsPublic/blob/master/devicetypes/zenwithin/zen-thermostat.src/zen-thermostat.groovy

SmartThingsPublic/blob/master/devicetypes/smartthings/zwave-thermostat.src/zwave-thermostat.groovy

SmartThingsPublic/blob/master/devicetypes/smartthings/ct100-thermostat.src/ct100-thermostat.groovy

**Comment by imnotbob** [comment]
The challenge in this is Nest does not report fan-only as a state.  It only reports
heating, cooling, idle.

On the thermostat setpoints, nest does not provide the other setpoints.

In both cases, "something" would have to be made up, or guessed, as the states doe not exist in Nest.   We have chosen not to guess states that don't exist, as the guess is impossible to validate what % of the time it is right or wrong.

Nests can be configured on to heating only or cooling only configs, and the other points do not exist.

In short, we cannot report things we nor nest "know".    I suggest for fan use use fanoperating state (which nest reports).   I suggest also that you know if systems are heating only and cooling only and don't request setpoints which don't exist.   

I do see the challenge in all this for both of us, but a limitation that ST asks devices to try to guess things, and we have no data on which to show a guess is at all reliable, and have folks make decisions on that guess.

I appreciate other thermostats may operate differently.   In some cases they may know what they are sending is accurate, and in other cases they are making things up that don't exist in the HVAC system.

**Comment by yracine** [comment]
Hi,

Ok, I understand that there are some limitations in the Nest APIs.

But, I think that you should reconsider at least the issues with the setpoints.  

Could you at least report the last heating or cooling setpoint and not obfuscate them?

The way I see it, your educated guess is better than mine (which is not as educated as I'm an outsider).

At the moment, because of the fact that you nullify the setpoints, I have to provide my own default values (which are not in any way as good as yours).

**P.S. I will not do any special coding for the NST manager DTH as my smartapps handle  all thermostat Types and models. I'm just following the standard DTH capabilities that are common to all Z-Wave, Zigbee, Wi-Fi connected thermostats under ST.**

My 2 cents.

**Comment by imnotbob** [comment]
the challenge is we don't know them.  Nest has schedules, that we are not privy to.

I think it is a mistake to assume there is a heating or cooling setpoint when there is none.   I agree guessing is bad - both for us and for you.    In most cases in the data we see, we don't have a setpoint.  Nothing to remember, nothing to go back to.  If the system does not have one, why would any of us guess one?

I think other thermostats have these issues.  I don't know what they "make up", but we view "making it up" is mostly wrong.

**Comment by yracine** [comment]
BTW,  _fanoperating_ state is not a ST standard attribute as well. 

Could you report the info in _fanOperatingState_ to _thermostatOperatingState_ to be standard like all the other thermostats?

I'm just trying here to convince you to be more standard....

Regards.

**Comment by imnotbob** [comment]
The challenge I see is Nest does not report it to us that way.  I don't know what is gained / lost when we overrride a report we have from nest (idle, cooling, heating) with something they did not report.   I.e. we are guessing how their internal code works, and that it is ok to overrride what they tell us for state.

Nest does report fan operation separately and we report that in thermostatFanMode.

We have taken the attitude of maintaining nest reported states, as we do not know their interactions / overrides to be able "know better than Nest" what the state is.

I would think staying standard in a smartapp, you would use thermostatFanMode.  If there is no fan on the hvac system, it is not accurate, but when hasFan is set, it is accurate.

As any fyi, when a nest is deployed, it reports to us (and we report in custom attributes)  canHeat, canCool, hasFan which says if the hvac system presented controls/status to nest for these functions.

The issue we have is we have 16K+ devices, and we don't want to be guessing on states, as with this number of devices our guesses will be wrong.  So we "don't guess".   We report what Nest tells us.

**Comment by yracine** [comment]
Hi, 

I'm just trying here to better  understand the way you implemented your DTH, are you telling me that you're reporting the _fanOperatingState_ in _thermostatFanMode_?

For example, if only the fan is running (no heating or cooling at the same time), where do you report this event?  Is it in _thermostatFanMode_?

If so, you're definivetely not standard. The  _thermostatFanMode_ attribute is strictly used (as its name indicates) to switch fan modes (ex. from 'auto' to 'on' or from 'off' to 'circulate')...

_thermostatOperatingState_ is used to report any real time events happening at the thermostat....So, you just need to report the 'fan only' event in _thermostatOperatingState_ instead of _thermostatFanMode_ (if it's what you're doing now) when the _hasFan_ is set in your DTH.  

You'll then follow the standard ST way of reporting real time events which is followed by all the ST connected thermostats out there....

Imagine, if all the thermostat DTH authors have different ways to implement the _thermostatFanMode_  & _thermostatOperatingState_ capabilities, no smartapps will be able to work properly.

Again, I'm just trying here to convince you to be more standard. 

To ask the smartapp authors to shoulder the burden to determine if the fan is running  when there should be a way to do it in the DTH and following the ST standards is really not OK (in my book).

Regards.

**Comment by imnotbob** [comment]
I misspoke on fanOperatingState   we only report thermostatFanMode

I think you are "interpreting" standard - i.e. things not written down.   you can read and write thermostatFanMode.

I understand your point of view.   I don't see any way to judge it is correct 100% or even judge what % it is correct.

Some thermostats have mode as off/heat/cool  others have off/heat/cool/auto,

Usually fan on thermostats is a separate control.   You can force the fan on while the system is still
in mode heat/cool/auto.

So standard here is in the eye of the beholder.   My guess (I don't work for Nest), is they decide it is two controls that can be heat, with fan forced on.    I'm going by - report what nest reports, not trying to understand their internal logic.

I think you can be standard and use thermostatFanMode.   It will tell you on/off, and let you set the system on/off.

In short, you are asking for a guess and we don't want to guess on states presented to us.

We have no way to know or test that a guess is right a 100% of the time.  And we have seen numerous times where things don't work as one might think.

**Comment by yracine** [comment]
Hi, please just answer the following questions:

For example, if only the fan is running (no heating or cooling at the same time), where do you report this real time event? 

Is it in _thermostatFanMode_ or there is no way for you to report this real time event based on the existing NEST APIs?

Regards.

**Comment by imnotbob** [comment]
thermostatFanMode

**Comment by yracine** [comment]
OK, if so, you're able to detect when only the fan is running at any time.
 
If you don't want to change the existing code because you're afraid of the impact on your existing Nest customer base, could you also report this real time event in _thermostatOperatingState_?

This way, you'll be standard without fear of breaking anything elsewhere.

I know that you have a large customer base and you don't want to rock the boat.

The solution I'm offering you would solve Gabriel's issue (and others) regarding the 'fan only' event.

Regards.

**Comment by DarellCraighead** [comment]
Well this discussion is certainly a challenge to keep up with.  I too have a Nest and am using ScheduleTstatZones as well.  It would be nice if the only known Nest DTH for ST would fill that gap.

imnotbob, I am curious about what you said here:

> The challenge in this is Nest does not report fan-only as a state. It only reports
> heating, cooling, idle.

So what does Nest report if it is in cool mode and I hit the fan button and tell it to run 15 minutes while cooling is idle?

I do see your point here that Nest does not have a fan on/auto setting that most tstats do...  that might be key.

> Nests can be configured on to heating only or cooling only configs, and the other points do not exist.

Not sure why you said this as Nest has Heat, Cool, Heat-Cool (aka Auto), Eco, and off as possible modes.  If it is in heat-cool mode then there are setpoints for both heat and cool.

**Comment by imnotbob** [comment]
Nest reports the fan auto/on in the thermostatFanMode.   So if you hit the button you will see it change  that change.  

Nest changes what it reports based on both modes, and how the thermostat is installed.  it "knows" based on the installation / wiring, what it can/cannot do.   So what modes or setpoint are available are not constant, but change based on the wiring.

The way to control using St standard capabilities is to use thermostatFanMode.  This will tell you if the fan is on or not, and let you set it on, if some automation requests it to be on.

I think relying on data that a thermostat has a setpoint when it does not, is a problem in building an automation.   The data for the decision does not exist.   That some thermostats "make it up" does not change the issue that the data does not exist and the thermostat does not control that function in this particular installation/configuration.  Remembering something old does not work when there was never anything to remember.

**Comment by yracine** [comment]
Hi again, 

When you say that "Nest reports the fan auto/on in the thermostatFanMode. So if you hit the button you will see it change that change", this is the behavior that you implemented in your DTH, not Nest.

So, to recap, what the Nest Users are asking here is that you also report the 'fan only' event in _thermostatOperatingState_ according to the ST standard capabilities.

The same principle applies for the thermostat's setpoints.  You have plenty of state variables in your DTH and some of them should keep the last cooling/heating setpoints for that purpose, and you should not nullify any setpoints.

I understand that the last item seems to be a problem for you guys based on the NEST APIs available, but in order for other ST developers to use your DTH in a standard way, you need to follow the ST guidelines for all thermostat capabilities.  We all have different ST challenges, and I'm pretty confident that with the brain power you have as a team that supports this **great DTH,** you'll be able to tackle all those different issues.

I will not repeat myself anymore as I feel that this conversation is going round in circles...

Good luck in your endeavours.

**Comment by tonesto7** [comment]
This has been added into our upcoming release.  I will likely be pushing it out tonight or tomorrow

**Comment by tonesto7** [comment]
@gaboferrari Can you verify this is working for you as intended now?
